### DISCUSSION TITLE
Trying to reproduce hg+Phabricator from Facebook but for git+GitHub

### DISCUSSION THREAD

**Discussion by mlanln** [arxanas/git-branchless#860] [Q&A] [answered]
I loved using Mercurial at Facebook, combined with `jf` for easily creating Phabricator diffs from stacks. I'm trying to get this experience with `git-branchless` + `ghstack` + GitHub.

I found git-branchless is a great option for the source control management component - smartlog, creating stacks, rebasing, navigating between commits is easy.

To that end, I have a few questions as I figure out how this works:

## question 1: `git sl`

If I do a clone of https://github.com/mlanln/github-testing (which only has one commit) and create 2 local commits, I expect the smartlog to show the public commit and the 2 local commits. But it only shows the latest local commit (see last `git sl` output):

```
% git sl
◆ d2fcd62 2h (ᐅ main) Initial commit
% echo 'change #1' > README.md
% git record -m 'change #1'
branchless: processing 2 updates: branch main, ref HEAD
branchless: processed commit: 2b4e6d6 change #1
[main 2b4e6d6] change #1
 1 file changed, 1 insertion(+), 1 deletion(-)
% git sl
⋮
◆ 2b4e6d6 9s (ᐅ main) change #1
% echo 'change #2' > README.md
% git record -m 'change #2'
branchless: processing 2 updates: branch main, ref HEAD
branchless: processed commit: 7e0a221 change #2
[main 7e0a221] change #2
 1 file changed, 1 insertion(+), 1 deletion(-)
% git sl
⋮
◆ 7e0a221 6s (ᐅ main) change #2
```

I expect it to show:

```
◆ d2fcd62 2h (ᐅ main) Initial commit
|
◆ 2b4e6d6 9s (ᐅ main) change #1
|
◆ 7e0a221 6s (ᐅ main) change #2
```

Any idea why this is?

## question 2: ghstack interaction

Edit: Never mind, please ignore question 2. I think I need to spend more time playing around with it first.

~~For the diffs / pull request side of it, the most promising option I've tried is [ghstack](https://github.com/ezyang/ghstack). It creates a separate pull request for each commit in the stack and records the stack of pull requests in the description of each pull request. However, it seems to not play nice with git-branchless.~~

~~For example, continuing on from the previous repo state, I run `ghstack`, and it correctly creates two pull requests. But it also edits the commit message with some metadata like the Pull Request, creating duplicate commits.~~

~~I know I could make a wrapper script to simply `git hide` the old commits, but wondering if there is a better way to handle it.~~

**Comment by arxanas** [comment, marked as answer]
Hi @mlanln,

> If I do a clone of https://github.com/mlanln/github-testing (which only has one commit) and create 2 local commits, I expect the smartlog to show the public commit and the 2 local commits. But it only shows the latest local commit (see last git sl output):

git-branchless doesn't have a real equivalent to Mercurial's notion of commit "phases", so instead, it defines "public" commits to be all commits which are ancestors of the main branch, and "draft" commits to be all other visible commits (see https://github.com/arxanas/git-branchless/wiki/Command:-git-smartlog#which-commits-are-visible). The indicator `(main)` would mean that the branch `main` points to this commit, and the `ᐅ` in `(ᐅ main)` means that `main` is checked out.

In your case, you are committing to the local `main` branch, which is causing the local commits to be treated as public. You probably want to detach the branch before committing, such as with `git sw -d`, or detach as part of committing with `git record -d ...`. This might take some time to get used to. You can also configure this: https://github.com/arxanas/git-branchless/wiki/Configuration#branchlessnavigationautoswitchbranches
